### ISSUE TITLE
Bug: False Negative `no-redeclare`

### ISSUE THREAD

**Issue by Yash-Singh1** [eslint/eslint#16137]
### Environment

Node version: N/A 
npm version: N/A 
Local ESLint version: N/A 
Global ESLint version: N/A 
Operating System: N/A 

See playground link

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```json
{
    "parserOptions": {
        "ecmaVersion": 12,
        "sourceType": "script",
        "ecmaFeatures": {}
    },
    "rules": {
        "no-redeclare": [
            "error"
        ]
    },
    "env": {}
}
```
</details>

```js
function a() { 
  return null
}
if (true) {
  let a = true
}
```

https://eslint.org/play#eyJ0ZXh0IjoiZnVuY3Rpb24gYSgpIHsgXG4gIHJldHVybiBudWxsXG59XG5pZiAodHJ1ZSkge1xuICBsZXQgYSA9IHRydWVcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXJlZGVjbGFyZSI6WyJlcnJvciJdfSwiZW52Ijp7fX19

### What did you expect to happen?

`no-redeclare` to report on the redeclaration of `a` on Line 5.

### What actually happened?

Empty

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I'm not sure if this is expected behavior, but variables that are currently declared are not considered as declared by the rule in child scopes, so when the rule is iterating over the if statement's scope it doesn't know that `a` even exists.

**Comment by mdjermanovic** [comment]
This works as intended. The `no-redeclare` rule reports multiple declarations of the same variable. In this case, inner `a` (`let a`) and outer `a` (`function a`) are two different variables, so this is not a case with multiple declarations of the same variable. This is a case that would be reported by the [no-shadow](https://eslint.org/docs/latest/rules/no-shadow) rule:

https://eslint.org/play#eyJ0ZXh0IjoiZnVuY3Rpb24gYSgpIHsgXG4gIHJldHVybiBudWxsXG59XG5pZiAodHJ1ZSkge1xuICBsZXQgYSA9IHRydWVcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjEyLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLXNoYWRvdyI6WyJlcnJvciJdfSwiZW52Ijp7fX19
